lib/Dialect/OM/Transforms/LinkModules.cpp

```diff
@@ -106,6 +134,96 @@ void ModuleInfo::postProcess(const SymMappingTy &symMapping) {
   });
 }
 
+// Return a failure if modules cannot be resolved. Return true if
+// it's necessary to rename symbols.
+static FailureOr<bool> resolveHWModules(StringAttr name,
+                                        ArrayRef<HWModuleLike> hwModules) {
+  bool existsExternalModule = false;
+  size_t countDefinition = 0;
+  HWModuleOp hwModule;
+
+  bool isPrivate = false;
+  SmallVector<Location> publicModules;
+  for (auto op : hwModules) {
+    if (isa<HWModuleExternOp>(op))
+      existsExternalModule = true;
+    else if (!countDefinition) {
+      hwModule = cast<HWModuleOp>(op);
+      isPrivate = hwModule.isPrivate();
+      ++countDefinition;
+    } else if (hwModule.isPrivate() && isPrivate) {
+      ++countDefinition;
+    } else {
+      publicModules.push_back(op->getLoc());
+    }
+  }
```

Comment:
Why it's different from `resolveClasses`?

Reply:
Few differences like the public modules shouldn't be renamed on conflict, and a simple check for module ports equality between extern modules.